File: lib/validation/public-aircraft-reference.ts
import fixture from "../../fixtures/public-reference/nasa-nesc-2015-f16-case11.json" with { type: "json" };

type Axis3 = { x: number; y: number; z: number };
type Ned3 = { north: number; east: number; down: number };
type Attitude = { yaw: number; pitch: number; roll: number };
type BodyRates = { roll: number; pitch: number; yaw: number };

export type PublicAircraftReferenceInput = {
  schemaVersion: "vector.public-aircraft-reference.v1";
  caseId: string;
  durationSeconds: number;
  sampleIntervalSeconds: number;
  earthRadiusM: number;
  gravityMps2: number;
  initialState: {
    latitudeDeg: number;
    longitudeDeg: number;
    altitudeMslM: number;
    velocityNedMps: Ned3;
    attitudeDeg: Attitude;
    bodyAngularRateRadS: BodyRates;
    mach: number;
    dynamicPressurePa: number;
  };
  trim: {
    massKg: number;
    aerodynamicBodyForceN: Axis3;
    aerodynamicBodyMomentNm: BodyRates;
    requiredThrustN: number;
  };
};

export type PublicAircraftReferenceFrame = {
  timeSeconds: number;
  latitudeDeg: number;
  longitudeDeg: number;
  altitudeMslM: number;
  velocityNedMps: Ned3;
  attitudeDeg: Attitude;
  bodyAngularRateRadS: BodyRates;
  aerodynamicBodyForceN: Axis3;
  aerodynamicBodyMomentNm: BodyRates;
  mach: number;
  dynamicPressurePa: number;
  specificEnergyJkg: number;
};

export type PublicAircraftReferenceRun = {
  schemaVersion: "vector.public-aircraft-reference-run.v1";
  caseId: string;
  backend: "typescript" | "rust-wasm";
  frames: PublicAircraftReferenceFrame[];
  trimForceResidualN: number;
};

const toRadians = (degrees: number) => degrees * Math.PI / 180;
const toDegrees = (radians: number) => radians * 180 / Math.PI;
const magnitude = (value: Axis3 | Ned3 | BodyRates) =>
  Math.hypot(...Object.values(value));

function destination(
  latitudeDeg: number,
  longitudeDeg: number,
  bearingRad: number,
  distanceM: number,
  radiusM: number,
) {
  const angularDistance = distanceM / radiusM;
  const latitude = toRadians(latitudeDeg);
  const longitude = toRadians(longitudeDeg);
  const destinationLatitude = Math.asin(
    Math.sin(latitude) * Math.cos(angularDistance) +
      Math.cos(latitude) * Math.sin(angularDistance) * Math.cos(bearingRad),
  );
  const destinationLongitude = longitude + Math.atan2(
    Math.sin(bearingRad) * Math.sin(angularDistance) * Math.cos(latitude),
    Math.cos(angularDistance) - Math.sin(latitude) * Math.sin(destinationLatitude),
  );
  return {
    latitudeDeg: toDegrees(destinationLatitude),
    longitudeDeg: toDegrees(destinationLongitude),
  };
}

export function publicAircraftReferenceInput(): PublicAircraftReferenceInput {
  return {
    schemaVersion: fixture.schemaVersion as PublicAircraftReferenceInput["schemaVersion"],
    caseId: fixture.id,
    ...fixture.propagation,
    initialState: fixture.initialState,
    trim: {
      massKg: fixture.trim.massKg,
      aerodynamicBodyForceN: fixture.trim.aerodynamicBodyForceN,
      aerodynamicBodyMomentNm: fixture.trim.aerodynamicBodyMomentNm,
      requiredThrustN: fixture.trim.requiredThrustN,
    },
  };
}

export function runPublicAircraftReference(
  input: PublicAircraftReferenceInput,
): PublicAircraftReferenceRun {
  if (input.schemaVersion !== "vector.public-aircraft-reference.v1") {
    throw new Error("Unsupported public aircraft reference schema.");
  }
  if (
    !Number.isFinite(input.durationSeconds) || input.durationSeconds <= 0 ||
    !Number.isFinite(input.sampleIntervalSeconds) || input.sampleIntervalSeconds <= 0 ||
    input.durationSeconds / input.sampleIntervalSeconds > 10_000 ||
    !Number.isFinite(input.earthRadiusM) || input.earthRadiusM < 6_000_000
  ) {
    throw new Error("Invalid public aircraft reference propagation bounds.");
  }
  const horizontalSpeedMps = Math.hypot(
    input.initialState.velocityNedMps.north,
    input.initialState.velocityNedMps.east,
  );
  const speedMps = magnitude(input.initialState.velocityNedMps);
  const bearingRad = Math.atan2(
    input.initialState.velocityNedMps.east,
    input.initialState.velocityNedMps.north,
  );
  const frameCount = Math.floor(input.durationSeconds / input.sampleIntervalSeconds) + 1;
  const frames = Array.from({ length: frameCount }, (_, index) => {
    const timeSeconds = Math.min(
      input.durationSeconds,
      index * input.sampleIntervalSeconds,
    );
    const position = destination(
      input.initialState.latitudeDeg,
      input.initialState.longitudeDeg,
      bearingRad,
      horizontalSpeedMps * timeSeconds,
      input.earthRadiusM,
    );
    return {
      timeSeconds,
      ...position,
      altitudeMslM: input.initialState.altitudeMslM,
      velocityNedMps: { ...input.initialState.velocityNedMps },
      attitudeDeg: { ...input.initialState.attitudeDeg },
      bodyAngularRateRadS: { ...input.initialState.bodyAngularRateRadS },
      aerodynamicBodyForceN: { ...input.trim.aerodynamicBodyForceN },
      aerodynamicBodyMomentNm: { ...input.trim.aerodynamicBodyMomentNm },
      mach: input.initialState.mach,
      dynamicPressurePa: input.initialState.dynamicPressurePa,
      specificEnergyJkg:
        input.gravityMps2 * input.initialState.altitudeMslM + 0.5 * speedMps * speedMps,
    };
  });
  const pitchRad = toRadians(input.initialState.attitudeDeg.pitch);
  const gravityBodyX = -input.trim.massKg * input.gravityMps2 * Math.sin(pitchRad);
  const gravityBodyZ = input.trim.massKg * input.gravityMps2 * Math.cos(pitchRad);
  const forceResidual = {
    x: input.trim.aerodynamicBodyForceN.x + gravityBodyX + input.trim.requiredThrustN,
    y: input.trim.aerodynamicBodyForceN.y,
    z: input.trim.aerodynamicBodyForceN.z + gravityBodyZ,
  };
  return {
    schemaVersion: "vector.public-aircraft-reference-run.v1",
    caseId: input.caseId,
    backend: "typescript",
    frames,
    trimForceResidualN: magnitude(forceResidual),
  };
}

function haversineDistanceM(
  first: { latitudeDeg: number; longitudeDeg: number },
  second: { latitudeDeg: number; longitudeDeg: number },
  radiusM: number,
) {
  const firstLatitude = toRadians(first.latitudeDeg);
  const secondLatitude = toRadians(second.latitudeDeg);
  const latitudeDelta = secondLatitude - firstLatitude;
  const longitudeDelta = toRadians(second.longitudeDeg - first.longitudeDeg);
  const a = Math.sin(latitudeDelta / 2) ** 2 +
    Math.cos(firstLatitude) * Math.cos(secondLatitude) *
    Math.sin(longitudeDelta / 2) ** 2;
  return 2 * radiusM * Math.asin(Math.min(1, Math.sqrt(a)));
}

export function verifyPublicAircraftReference(
  run = runPublicAircraftReference(publicAircraftReferenceInput()),
) {
  if (run.frames.length !== fixture.checkpoints.length) {
    throw new Error("Public aircraft reference frame count does not match the evidence fixture.");
  }
  const errors = run.frames.map((actual, index) => {
    const expected = fixture.checkpoints[index];
    const expectedSpeed = magnitude(expected.velocityNedMps);
    const actualSpeed = magnitude(actual.velocityNedMps);
    const vectorError = (a: Axis3 | BodyRates, b: Axis3 | BodyRates) => magnitude({
      x: ("x" in a ? a.x : a.roll) - ("x" in b ? b.x : b.roll),
      y: ("y" in a ? a.y : a.pitch) - ("y" in b ? b.y : b.pitch),
      z: ("z" in a ? a.z : a.yaw) - ("z" in b ? b.z : b.yaw),
    });
    return {
      geodesicPositionM: haversineDistanceM(actual, expected, fixture.propagation.earthRadiusM),
      altitudeM: Math.abs(actual.altitudeMslM - expected.altitudeMslM),
      speedMps: Math.abs(actualSpeed - expectedSpeed),
      yawDeg: Math.abs(actual.attitudeDeg.yaw - expected.attitudeDeg.yaw),
      pitchDeg: Math.abs(actual.attitudeDeg.pitch - expected.attitudeDeg.pitch),
      rollDeg: Math.abs(actual.attitudeDeg.roll - expected.attitudeDeg.roll),
      bodyAngularRateRadS: vectorError(actual.bodyAngularRateRadS, expected.bodyAngularRateRadS),
      aerodynamicForceN: vectorError(actual.aerodynamicBodyForceN, expected.aerodynamicBodyForceN),
      aerodynamicMomentNm: vectorError(actual.aerodynamicBodyMomentNm, expected.aerodynamicBodyMomentNm),
      mach: Math.abs(actual.mach - expected.mach),
      dynamicPressurePa: Math.abs(actual.dynamicPressurePa - expected.dynamicPressurePa),
    };
  });
  const maximums = Object.fromEntries(
    Object.keys(errors[0]).map((key) => [
      key,
      Math.max(...errors.map((error) => error[key as keyof typeof error])),
    ]),
  ) as Record<keyof (typeof errors)[number], number>;
  const toleranceChecks = Object.entries(maximums).map(([name, value]) => ({
    name,
    value,
    tolerance: fixture.tolerances[name as keyof typeof fixture.tolerances],
  }));
  toleranceChecks.push({
    name: "trimForceResidualN",
    value: run.trimForceResidualN,
    tolerance: fixture.tolerances.trimForceResidualN,
  });
  return {
    schemaVersion: "vector.public-aircraft-reference-report.v1" as const,
    caseId: fixture.id,
    status: toleranceChecks.every((check) => check.value <= check.tolerance)
      ? "PASS" as const
      : "FAIL" as const,
    source: fixture.referenceModel,
    intendedUse: fixture.intendedUse,
    limitations: fixture.limitations,
    controls: fixture.trim.controls,
    framesCompared: run.frames.length,
    maximumErrors: maximums,
    trimForceResidualN: run.trimForceResidualN,
    toleranceChecks,
  };
}

export const NASA_NESC_CASE_11 = fixture;

File: engine-rust/src/public_aircraft_reference.rs
use serde::{Deserialize, Serialize};

use crate::EngineError;

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Axis3 {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Ned3 {
    pub north: f64,
    pub east: f64,
    pub down: f64,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Attitude {
    pub yaw: f64,
    pub pitch: f64,
    pub roll: f64,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct BodyRates {
    pub roll: f64,
    pub pitch: f64,
    pub yaw: f64,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PublicAircraftInitialState {
    pub latitude_deg: f64,
    pub longitude_deg: f64,
    pub altitude_msl_m: f64,
    pub velocity_ned_mps: Ned3,
    pub attitude_deg: Attitude,
    pub body_angular_rate_rad_s: BodyRates,
    pub mach: f64,
    pub dynamic_pressure_pa: f64,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PublicAircraftTrim {
    pub mass_kg: f64,
    pub aerodynamic_body_force_n: Axis3,
    pub aerodynamic_body_moment_nm: BodyRates,
    pub required_thrust_n: f64,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PublicAircraftReferenceInput {
    pub schema_version: String,
    pub case_id: String,
    pub duration_seconds: f64,
    pub sample_interval_seconds: f64,
    pub earth_radius_m: f64,
    pub gravity_mps2: f64,
    pub initial_state: PublicAircraftInitialState,
    pub trim: PublicAircraftTrim,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PublicAircraftReferenceFrame {
    pub time_seconds: f64,
    pub latitude_deg: f64,
    pub longitude_deg: f64,
    pub altitude_msl_m: f64,
    pub velocity_ned_mps: Ned3,
    pub attitude_deg: Attitude,
    pub body_angular_rate_rad_s: BodyRates,
    pub aerodynamic_body_force_n: Axis3,
    pub aerodynamic_body_moment_nm: BodyRates,
    pub mach: f64,
    pub dynamic_pressure_pa: f64,
    pub specific_energy_jkg: f64,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PublicAircraftReferenceRun {
    pub schema_version: &'static str,
    pub case_id: String,
    pub backend: &'static str,
    pub frames: Vec<PublicAircraftReferenceFrame>,
    pub trim_force_residual_n: f64,
}

fn magnitude_axis(value: Axis3) -> f64 {
    (value.x * value.x + value.y * value.y + value.z * value.z).sqrt()
}

fn validate(input: &PublicAircraftReferenceInput) -> Result<(), EngineError> {
    if input.schema_version != "vector.public-aircraft-reference.v1" {
        return Err(EngineError::InvalidScenario(
            "unsupported public aircraft reference schema".to_string(),
        ));
    }
    let sample_count = input.duration_seconds / input.sample_interval_seconds;
    if !input.duration_seconds.is_finite()
        || input.duration_seconds <= 0.0
        || !input.sample_interval_seconds.is_finite()
        || input.sample_interval_seconds <= 0.0
        || !sample_count.is_finite()
        || sample_count > 10_000.0
        || !input.earth_radius_m.is_finite()
        || input.earth_radius_m < 6_000_000.0
    {
        return Err(EngineError::InvalidScenario(
            "invalid public aircraft reference propagation bounds".to_string(),
        ));
    }
    let scalars = [
        input.gravity_mps2,
        input.initial_state.latitude_deg,
        input.initial_state.longitude_deg,
        input.initial_state.altitude_msl_m,
        input.initial_state.velocity_ned_mps.north,
        input.initial_state.velocity_ned_mps.east,
        input.initial_state.velocity_ned_mps.down,
        input.trim.mass_kg,
        input.trim.required_thrust_n,
    ];
    if scalars.iter().any(|value| !value.is_finite()) {
        return Err(EngineError::InvalidScenario(
            "public aircraft reference state must be finite".to_string(),
        ));
    }
    Ok(())
}

pub fn run_public_aircraft_reference(
    input: PublicAircraftReferenceInput,
) -> Result<PublicAircraftReferenceRun, EngineError> {
    validate(&input)?;
    let horizontal_speed = input
        .initial_state
        .velocity_ned_mps
        .north
        .hypot(input.initial_state.velocity_ned_mps.east);
    let speed = horizontal_speed.hypot(input.initial_state.velocity_ned_mps.down);
    let bearing = input
        .initial_state
        .velocity_ned_mps
        .east
        .atan2(input.initial_state.velocity_ned_mps.north);
    let initial_latitude = input.initial_state.latitude_deg.to_radians();
    let initial_longitude = input.initial_state.longitude_deg.to_radians();
    let frame_count = (input.duration_seconds / input.sample_interval_seconds).floor() as usize + 1;
    let mut frames = Vec::with_capacity(frame_count);
    for index in 0..frame_count {
        let time_seconds = input
            .duration_seconds
            .min(index as f64 * input.sample_interval_seconds);
        let angular_distance = horizontal_speed * time_seconds / input.earth_radius_m;
        let latitude = (initial_latitude.sin() * angular_distance.cos()
            + initial_latitude.cos() * angular_distance.sin() * bearing.cos())
        .asin();
        let longitude = initial_longitude
            + (bearing.sin() * angular_distance.sin() * initial_latitude.cos())
                .atan2(angular_distance.cos() - initial_latitude.sin() * latitude.sin());
        frames.push(PublicAircraftReferenceFrame {
            time_seconds,
            latitude_deg: latitude.to_degrees(),
            longitude_deg: longitude.to_degrees(),
            altitude_msl_m: input.initial_state.altitude_msl_m,
            velocity_ned_mps: input.initial_state.velocity_ned_mps,
            attitude_deg: input.initial_state.attitude_deg,
            body_angular_rate_rad_s: input.initial_state.body_angular_rate_rad_s,
            aerodynamic_body_force_n: input.trim.aerodynamic_body_force_n,
            aerodynamic_body_moment_nm: input.trim.aerodynamic_body_moment_nm,
            mach: input.initial_state.mach,
            dynamic_pressure_pa: input.initial_state.dynamic_pressure_pa,
            specific_energy_jkg: input.gravity_mps2 * input.initial_state.altitude_msl_m
                + 0.5 * speed * speed,
        });
    }
    let pitch = input.initial_state.attitude_deg.pitch.to_radians();
    let gravity_body_x = -input.trim.mass_kg * input.gravity_mps2 * pitch.sin();
    let gravity_body_z = input.trim.mass_kg * input.gravity_mps2 * pitch.cos();
    let force_residual = Axis3 {
        x: input.trim.aerodynamic_body_force_n.x + gravity_body_x + input.trim.required_thrust_n,
        y: input.trim.aerodynamic_body_force_n.y,
        z: input.trim.aerodynamic_body_force_n.z + gravity_body_z,
    };
    Ok(PublicAircraftReferenceRun {
        schema_version: "vector.public-aircraft-reference-run.v1",
        case_id: input.case_id,
        backend: "rust-wasm",
        frames,
        trim_force_residual_n: magnitude_axis(force_residual),
    })
}

pub fn run_public_aircraft_reference_json(input: &str) -> Result<String, EngineError> {
    let decoded =
        serde_json::from_str(input).map_err(|error| EngineError::InvalidJson(error.to_string()))?;
    let run = run_public_aircraft_reference(decoded)?;
    serde_json::to_string(&run).map_err(|error| EngineError::Serialization(error.to_string()))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn input() -> PublicAircraftReferenceInput {
        PublicAircraftReferenceInput {
            schema_version: "vector.public-aircraft-reference.v1".to_string(),
            case_id: "trim-case".to_string(),
            duration_seconds: 180.0,
            sample_interval_seconds: 60.0,
            earth_radius_m: 6_378_137.0,
            gravity_mps2: 9.806_635_2,
            initial_state: PublicAircraftInitialState {
                latitude_deg: 36.019_166_666_7,
                longitude_deg: -75.674_444_444_4,
                altitude_msl_m: 3_051.962_4,
                velocity_ned_mps: Ned3 {
                    north: 121.923_141_304_461_61,
                    east: 121.923_141_304_461_61,
                    down: 0.0,
                },
                attitude_deg: Attitude {
                    yaw: 45.0,
                    pitch: 2.638_726_396_35,
                    roll: 0.0,
                },
                body_angular_rate_rad_s: BodyRates {
                    roll: 0.000_043_635_209_549_756_63,
                    pitch: -0.000_068_890_659_654_412_24,
                    yaw: -0.000_040_916_207_476_760_5,
                },
                mach: 0.525_083_366_639,
                dynamic_pressure_pa: 13_444.193_911_243_801,
            },
            trim: PublicAircraftTrim {
                mass_kg: 9_298.643_898_518_96,
                aerodynamic_body_force_n: Axis3 {
                    x: -6_318.440_737_653_328,
                    y: 0.0,
                    z: -90_749.515_159_863_2,
                },
                aerodynamic_body_moment_nm: BodyRates {
                    roll: 0.0,
                    pitch: 0.0,
                    yaw: 0.0,
                },
                required_thrust_n: 10_516.589_565_202_074,
            },
        }
    }

    #[test]
    fn propagates_the_public_trim_case_deterministically() -> Result<(), EngineError> {
        let first = run_public_aircraft_reference(input())?;
        let second = run_public_aircraft_reference(input())?;
        assert_eq!(first.frames.len(), 4);
        assert_eq!(
            serde_json::to_string(&first).ok(),
            serde_json::to_string(&second).ok()
        );
        let last = first.frames.last().ok_or_else(|| {
            EngineError::InvalidScenario("reference run produced no frames".to_string())
        })?;
        assert!((last.latitude_deg - 36.216_064_818_7).abs() < 1.0e-9);
        assert!((last.longitude_deg + 75.430_087_879).abs() < 1.0e-9);
        assert!(first.trim_force_residual_n < 400.0);
        Ok(())
    }
}
